### PR TITLE
Rail junction switch (GCI task)

### PR DESCRIPTION
**Pull request by anuar2k**
I just want to add that AdditionalRails module must be Signalling module dependent now.

Aleksander Wójtowicz

### CODE REVIEW

File: assets/blocks/railSwitch/railSwitchCombinedOff.block [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "entity": {
+        "prefab": "AdditionalRails:railSwitchCombinedOff"
+    }
+}
```
Comment (Cervator) [inline assets/blocks/railSwitch/railSwitchCombinedOff.block]: Whenever you see these weird little red symbols in diff it is an indicator of a missing newline at the end of a file, which is a good practice to follow. Some editors like Atom or IntelliJ will tend to insert one for you :-)

File: src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java [hunk 2]
```diff
@@ -1,74 +1,167 @@
...
 import org.terasology.registry.In;
-import org.terasology.signalling.components.SignalConsumerComponent;
 import org.terasology.signalling.components.SignalConsumerStatusComponent;
 import org.terasology.world.BlockEntityRegistry;
 import org.terasology.world.WorldProvider;
 import org.terasology.world.block.Block;
 import org.terasology.world.block.BlockComponent;
 import org.terasology.math.geom.Vector3i;
 import org.terasology.world.block.BlockManager;
+import org.terasology.world.block.family.BlockFamily;
```
Comment (Cervator) [inline src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java]: This and some other imports here and there in affected files appear to not be used. If you spot those (even if you didn't introduce them) please fix by simply removing them :-) A good IDE will show which are unused.

File: src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java [hunk 3]
```diff
@@ -1,74 +1,167 @@
...
+                if (input) {
+                    railSwitchSignalOn.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchSignalOn);
+                } else {
+                    railSwitchSignalOff.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchSignalOff);
+                }
+                break;
+
+            case 2: //TODO: railSwitch using both activate (E button) and Signalling mode
```
Comment (Cervator) [inline src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java]: If this is new code you may want to leave a TODO but not a ton of commented out code. Either reference somewhere for inspiration (copy pasting) in the future or add the details to an issue in the repo instead :-)

File: src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java [hunk 4]
```diff
@@ -1,74 +1,167 @@
...
-                if (!blockEntityRegistry.getBlockEntityAt(blockLocation).hasComponent(RailSwitchComponent.class)) {
-                    blockEntityRegistry.getBlockEntityAt(blockLocation).addOrSaveComponent(railSwitchComponent);
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class})
+    public void activateAction(ActivateEvent event, EntityRef entity) {
+        RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
+
+        updateSwitch(event, entity, railSwitchComponent, !railSwitchComponent.isOn, ActionSource.ACTIVATE);
+    }
+
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class, SignalConsumerStatusComponent.class})
```
Comment (Cervator) [inline src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java]: When introducing new methods (and classes, etc) if at all possible please try to add some javadoc. Bonus points for also adding it to existing undocumented stuff you're interacting with anyway, if you understand it well enough to doc :-)

**Comment by anuar2k** [comment]
Closing the pull request as I'm going to open another one from other branch.

**Pull request closed without merging**